### ISSUE TITLE
naming the `#[idle]` function "main" produces a compiler error

### ISSUE THREAD

**Issue by japaric** [rtfm-rs/cortex-m-rtfm#311]
## Steps to reproduce:

Modify the `examples/idle.rs` in this repository like this:

``` diff
     }

     #[idle]
-    fn idle(_: idle::Context) -> ! {
+    fn main(_: main::Context) -> ! {
         static mut X: u32 = 0;

         // Safe access to local `static mut` variable
```

Now the example fails to build:

``` console
$ cargo b --example idle --target thumbv7m-none-eabi
error[E0061]: this function takes 0 arguments but 2 arguments were supplied
  --> examples/idle.rs:19:8
   |
11 | #[rtfm::app(device = lm3s6965)]
   | -------------------------------
   | |
   | supplied 2 arguments
   | defined here
...
19 |     fn main(_: main::Context) -> ! {
   |        ^^^^ expected 0 arguments

error: aborting due to previous error
```

## The problem

If you look into the `rtfm-expansion.rs` file you'll see that the proc-macro produces a `#[no_mangle]` function called "main"; this name collides with the user provided one.

``` rust
fn main(main::Locals { X, .. }: main::Locals, _: main::Context) -> ! {
    // user code
}

const APP: () = {
    use lm3s6965 as _;
    #[no_mangle]
    unsafe extern "C" fn main() -> ! {
        // ..
        main(
            main::Locals::new(),
            main::Context::new(&rtfm::export::Priority::new(0)),
        )
    }
};
```

## The fix

Modify the `main` function inside `const APP` to have a random identifier (it won't be called by user code or proc-macro generated code) and use `export_name` instead of `no_mangle`. So, like this:

``` rust
const APP: () = {
    use lm3s6965 as _;
    #[export_name = "main"]
    unsafe extern "C" fn __rtfm_0() -> ! {
        // ..
        main(
            main::Locals::new(),
            main::Context::new(&rtfm::export::Priority::new(0)),
        )
    }
};
```

Or change the inner `main` function call to use an absolute path: `crate::main(/* args */)`. That may also work.

**Comment by japaric** [comment]
The reason one may want to name the `#[idle]` function "main" is better integration with VS Code + Rust-Analyzer. RA gives you this "Run" button (which is sugar for `cargo run --example foo`) when it says a function named "main".

![2020-05-20-123040](https://user-images.githubusercontent.com/5018213/82437040-08681700-9a97-11ea-963c-e790601dac58.png)

**Comment by perlindgren** [comment]
With the new `mod`  scoping, maybe this will be solved by itself?

**Comment by japaric** [comment]
Possibly. I would still like to see this fixed in the v0.5.x series, which use `const APP`, though.

**Comment by perlindgren** [comment]
Right, this a bug. Should be fixed in 5.2. (And 6.0)
